File: lib/priority-engine.ts
import type { PatientVitals, Priority, ScoreBreakdown } from '@/types'
import { PRIORITY_THRESHOLDS } from '@/types'

export interface PriorityResult {
  urgencyScore: number
  priority: Priority
  breakdown: ScoreBreakdown[]
}

export function calculatePriority(
  vitals: PatientVitals,
  symptoms: string[],
  age: number
): PriorityResult {
  let score = 0
  const breakdown: ScoreBreakdown[] = []

  const addScore = (points: number, factor: string, condition: string) => {
    score += points
    breakdown.push({ factor, condition, points })
  }

  // SpO2
  const spo2 = vitals.spo2 || 0
  if (spo2 > 0) {
    if (spo2 < 85) addScore(40, 'SpO2', `${spo2}% — Severe Hypoxia`)
    else if (spo2 <= 89) addScore(25, 'SpO2', `${spo2}% — Moderate Hypoxia`)
    else if (spo2 <= 93) addScore(10, 'SpO2', `${spo2}% — Mild Hypoxia`)
  }

  // Blood Pressure
  const sys = vitals.bpSystolic || 0
  const dia = vitals.bpDiastolic || 0
  if (sys > 0 || dia > 0) {
    if (sys < 80 || (dia > 0 && dia < 50)) addScore(30, 'Blood Pressure', 'Critical Hypotension')
    else if (sys < 90) addScore(20, 'Blood Pressure', 'Hypotension')
    else if (sys > 180) addScore(15, 'Blood Pressure', 'Severe Hypertension')
    else if (sys > 160) addScore(8, 'Blood Pressure', 'Stage 2 Hypertension')
  }

  // Consciousness
  if (vitals.conscious === false) {
    addScore(35, 'Consciousness', 'Unconscious — Immediate priority')
  }

  // Heart Rate
  const hr = vitals.heartRate || 0
  if (hr > 0) {
    if (hr > 150 || hr < 40) addScore(20, 'Heart Rate', 'Critical Arrhythmia Risk')
    else if (hr > 120 || hr < 50) addScore(12, 'Heart Rate', 'Tachycardia / Bradycardia')
    else if (hr > 100) addScore(5, 'Heart Rate', 'Mild Tachycardia')
  }

  // Temperature
  const temp = vitals.temperature || 0
  if (temp > 0) {
    if (temp > 40.0 || temp < 35.0) addScore(15, 'Temperature', 'Extreme Temperature')
    else if (temp > 39.0) addScore(8, 'Temperature', 'High Fever')
    else if (temp < 36.0) addScore(8, 'Temperature', 'Hypothermia Risk')
  }

  // Symptoms
  const lowerSymptoms = symptoms.map(s => s.toLowerCase())
  const hasSymptom = (checks: string[]) => lowerSymptoms.some(s => checks.some(c => s.includes(c)))

  if (hasSymptom(['chest pain', 'chest pressure', 'chest tightness'])) addScore(15, 'Symptoms', 'Chest Pain')
  if (hasSymptom(['difficulty breathing', 'shortness of breath', 'dyspnea'])) addScore(15, 'Symptoms', 'Breathing Difficulty')
  if (hasSymptom(['stroke', 'facial droop', 'arm weakness', 'speech difficulty'])) addScore(25, 'Symptoms', 'Stroke Suspected')
  if (hasSymptom(['severe bleeding', 'hemorrhage'])) addScore(20, 'Symptoms', 'Severe Bleeding')
  if (hasSymptom(['seizure', 'convulsion'])) addScore(15, 'Symptoms', 'Seizures')
  if (hasSymptom(['anaphylaxis', 'throat swelling'])) addScore(25, 'Symptoms', 'Anaphylaxis')
  if (hasSymptom(['head injury', 'traumatic brain'])) addScore(15, 'Symptoms', 'Head Injury')

  // Age
  if (age < 2) addScore(10, 'Age', 'Infant')
  else if (age < 5) addScore(8, 'Age', 'Young child')
  else if (age > 80) addScore(10, 'Age', 'Elderly')
  else if (age > 70) addScore(5, 'Age', 'Senior')

  const urgencyScore = Math.min(100, score)

  let priority: Priority = 'P4'
  if (urgencyScore >= PRIORITY_THRESHOLDS.P1) priority = 'P1'
  else if (urgencyScore >= PRIORITY_THRESHOLDS.P2) priority = 'P2'
  else if (urgencyScore >= PRIORITY_THRESHOLDS.P3) priority = 'P3'

  return { urgencyScore, priority, breakdown }
}

File: app/services/priority_engine.py
def calculate_urgency(patient_data):
    """Calculate urgency score (0-100) and P1-P4 level"""
    score = 0
    reasons = []
    
    # Rule 1: Very low oxygen (< 90%)
    oxygen = patient_data.get('oxygen_level', 0)
    if oxygen < 90:
        score += 30
        reasons.append('Critical oxygen level (below 90%)')
    elif oxygen < 94:
        score += 15
        reasons.append('Low oxygen level (below 94%)')
    
    # Rule 2: Very low blood pressure (systolic < 90)
    systolic = patient_data.get('blood_pressure_systolic', 0)
    if systolic < 90:
        score += 30
        reasons.append('Critical low blood pressure')
    elif systolic < 100:
        score += 15
        reasons.append('Low blood pressure')
    
    # Rule 3: Unconscious patient
    if not patient_data.get('is_conscious', True):
        score += 40
        reasons.append('Patient unconscious')
    
    # Rule 4: Chest pain
    problem = patient_data.get('main_problem', '').lower()
    if 'chest' in problem:
        score += 20
        reasons.append('Chest pain reported')
    
    # Rule 5: Abnormal heart rate
    heart_rate = patient_data.get('heart_rate', 80)
    if heart_rate > 120:
        score += 15
        reasons.append('Tachycardia')
    elif heart_rate < 50:
        score += 15
        reasons.append('Bradycardia')
    
    # Rule 6: High fever (> 39°C)
    temp = patient_data.get('temperature', 37)
    if temp > 39:
        score += 10
        reasons.append('High fever')
    
    # Rule 7: Elderly (> 75 years)
    age = patient_data.get('age', 50)
    if age > 75:
        score += 10
        reasons.append('Elderly patient')
    
    # Cap at 100
    final_score = min(score, 100)
    
    # Determine priority level
    if final_score >= 80:
        level = 'P1'  # Critical
    elif final_score >= 60:
        level = 'P2'  # Urgent
    elif final_score >= 40:
        level = 'P3'  # Moderate
    else:
        level = 'P4'  # Stable
    
    return {
        'score': final_score,
        'level': level,
        'reasons': reasons
    }